### ISSUE TITLE
New Safe Types

### ISSUE THREAD

**Issue by Brains** [OfficeAutomata/NetMQ.High#1]
Initial ticket: https://github.com/OfficeAutomata/Dashboard/issues/151
As explained there I mostly added new types instead of changing existing ones to not break anything. Only expanded existing types to support inheritance of some functionality.

Usage of `ClientSafe`
```fs
   client <- new NetMQ.High.ClientSafe(address)
   client.Init() // Create Actor and call ClientSafeEngine.Initialize implicitly on a new thread
   client.Task // Await on Task to catch exceptions on Command thread and show them in UI
|> Async.AwaitTask
|> pairf (ignore >> o.Dispose |> tee >> unwrap) // If error then dispose everything created in client.Init
|> async.TryWith
|> Async.RunSynchronously // Run on Command thread
```

Usage of `ClientTimeout`
```fs
// create
client <- new NetMQ.High.ClientTimeout(address, 10000)
// rest is same with ClientSafe

// send
client.SendRequestAsync

// This method was removed so Client and ClientTimeout can be swapped in single line
// client.SendRequestAsyncWithTimeout 
```

Usage of `AsyncServerSafe`
```fs
   server <- new NetMQ.High.AsyncServerSafe(handler)
   server.Init()
   server.Bind address
   server.Task // Await on Task to catch exceptions on Command thread and show them in UI
|> Async.AwaitTask
|> pairf (ignore >> o.Dispose |> tee >> unwrap) // Dispose server if Bind was failed
|> async.TryWith
|> Async.RunSynchronously // Run on Command thread
```

Func `unwrap` is to unwrap `AggregateException` and expose its `InnerException`
```fs
let unwrap (e:System.Exception) = raise e.InnerException
```

**Comment by Brains** [comment]
> Yeah, NetMq has always been a pain

When tested yesterday new `Safe` types it turned out `NetMQ.High` is much worse. There were so many creepy bugs, crashes and faults that I was just totally lost among variety of them. They appeared during development and testing phases, I've fixed most and really hope all that horror won't appear in live version.
#
For example:
- Not proof against few `Dispose` calls while .NET Guides state a `Dispose` implementation should be
    - Plain `NetMQ` is not proof as well but errors appear on caller thread instead on a background which crashes the application
    - To handle the errors would need to override `ClientEngine.Cleanup` and pass `TaskCompletionSource` in the same as with `ClientEngine.Initialize`
    - So just added checks into `Datas.Manager.fs` code, much faster
- Calling `SendRequestAsync` blocks caller thread if `Client` is `Disposed`
    - No error on the caller or even background thread in this case, absolutely no hint what is wrong, total blindness
    - This block happens before `ClientTimeout.TimeoutAfter` call so `timeout` just does not work here and there is no way to make it working
    - It blocks the execution because `NetMQQueue.Enqueue` method _blocks if the queue is full_ (from docs) or `Disposed` (from my exploration)
- Not releases background threads properly so `Dashboard` process is still alive after closing main window
    - Thus, if connect-disconnect many times memory leaks are possible even without closing Dashboard
    - However, this was a problem with `NetMQ` as well

#

In conclusion, `NetMQ.High` is very brittle, totally unpredictable and not reliable at all. If more problems occur in future (highly probable) I suggest just get rid of it and follow [Actor](https://netmq.readthedocs.io/en/latest/actor/) pattern sample code. Luckily, we can migrate `Client` and `Server` independently, only need to preserve communication protocol, how they are implemented on each side should be less important.

**Comment by Brains** [comment]
Forgot to mention you can explore new tests to see how everything works:
- `Tests/AsyncServerSafeTests.cs`
- `Tests/ClientSafeDisposeTests.cs`
- `Tests/ClientSafeTests.cs`
- `Tests/ClientTimeoutServerTests.cs`
